prow/plugins/verify-owners/verify-owners.go

```diff
@@ -0,0 +1,136 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package verifyowners
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+	"k8s.io/test-infra/prow/repoowners"
+)
+
+const (
+	pluginName     = "verify-owners"
+	ownersFileName = "OWNERS"
+)
+
+var (
+	invalidOwnersLabel = "do-not-merge/invalid-OWNERS"
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	return &pluginhelp.PluginHelp{
+			Description: "The owners plugin validates formatting and 'labels' sections of OWNERS files modified in a PR. On failure it automatically adds a 'do-not-merge/invalid-OWNERS' label to the PR.",
+		},
+		nil
+}
+
+type ownersClient interface {
+	LoadRepoOwners(org, repo, base string) (repoowners.RepoOwnerInterface, error)
+}
+
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	CreateComment(owner, repo string, number int, comment string) error
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+}
+
+func handlePullRequest(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened && pre.Action != github.PullRequestActionReopened && pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+	return handle(pc.GitHubClient, pc.GitClient, pc.Logger, &pre, pc.PluginConfig.Owners.LabelsBlackList)
+}
+
+func handle(ghc githubClient, gc *git.Client, log *logrus.Entry, pre *github.PullRequestEvent, labelsBlackList []string) error {
+	org := pre.Repo.Owner.Login
+	repo := pre.Repo.Name
+	var errMsg error
+
+	// Get changes.
+	changes, err := ghc.GetPullRequestChanges(org, repo, pre.Number)
+	if err != nil {
+		return fmt.Errorf("error getting PR changes: %v", err)
+	}
+
+	// List modified OWNERS files.
+	var modifiedOwnersFiles []string
+	for _, change := range changes {
+		if filepath.Base(change.Filename) == ownersFileName {
+			modifiedOwnersFiles = append(modifiedOwnersFiles, change.Filename)
+		}
+	}
+	if len(modifiedOwnersFiles) == 0 {
+		return nil
+	}
+
+	// Clone the repo, checkout the PR.
+	r, err := gc.Clone(pre.Repo.FullName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Clean(); err != nil {
+			log.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+	if err := r.CheckoutPullRequest(pre.Number); err != nil {
+		return err
+	}
+
+	// Check each OWNERS file.
+	for _, f := range modifiedOwnersFiles {
+		// Try to load OWNERS file.
+		o := repoowners.InitOwners(r.Dir, false, nil, nil, log)
```

Comment:
There is no need to introduce these functions or load a RepoOwners struct at all. This plugin should just validate the changed OWNERS files. The `FindLabelsForFile`, and `Approvers` methods aren't needed (and aren't used in a valid way here).

Inside this loop unmarshal `f` and then validate the `labels` field and if it is the root OWNERS file also check that `approvers` is non empty.